src/crimson/os/seastore/cache.h

```diff
@@ -981,6 +985,7 @@ class Cache {
 
   // FIXME: This is specific to the segmented implementation
   SegmentProvider *segment_provider = nullptr;
+  SegmentProvider *cold_segment_provider = nullptr;
```

Comment:
Similar to `EPM::BackgroundProcess::cleaner_by_device_id`, we can instead introduce `std::vector<SegmentProvider *> segment_provider_by_device_id;`

So we can reuse `AsyncCleaner::get_device_ids()` instead of introducing another `SegmentProvider::contains()`.

The mechanism is temporary (see FIXME), and this is easier to drop later.

Reply:
How about

```cpp
void Cache::set_segment_provider(AsyncCleaner *cleaner) {
  auto provider = dynamic_cast<SegmentProvider*>(cleaner);
  if (provider) {
    for (auto id : cleaner->get_device_ids()) {
      segment_providers_by_device_id[id] = provider;
    }
  }
}
```
and
```cpp
const segment_info_t& Cache::get_segment_info(segment_id_t sid) {
  auto provider = segment_providers_by_device_id[sid.get_device_id()];
  assert(provider);
  return provider->get_seg_info(sid);
}
```